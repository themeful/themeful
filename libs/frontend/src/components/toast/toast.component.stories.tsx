import { h } from '@stencil/core'
import { timer } from 'rxjs'
import { mapTo } from 'rxjs/operators'

export default {
  title: 'Components/Toast',
}

export const toast = ({ first, second }): HTMLElement => {
  return (
    <div>
      <div class="header">
        <h1>Toast</h1>
      </div>
      <div class="content">
        <div style={{ position: 'relative', width: '400px', height: '100px' }}>
          <tf-toast {...first} />
        </div>
        <div style={{ position: 'relative', width: '400px', height: '100px' }}>
          <tf-toast {...second} />
        </div>
      </div>
    </div>
  )
}

toast.args = {
  first: {
    msg$: timer(0, 5000).pipe(
      mapTo({
        text: 'Hello World',
        status: 'success',
      })
    ),
  },
  second: {
    msg$: timer(0, 5000).pipe(
      mapTo({
        text: 'Hello World',
        status: 'error',
      })
    ),
  },
}
