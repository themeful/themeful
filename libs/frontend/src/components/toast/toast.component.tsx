import { Component, Event, EventEmitter, h, Host, Prop, State } from '@stencil/core'
import { Toast } from '@typings'
import { Observable, Subject, Subscription } from 'rxjs'
import { delay, filter, tap } from 'rxjs/operators'

@Component({
  tag: 'tf-toast',
  styleUrl: 'toast.component.scss',
  shadow: true,
})
export class ToastComponent {
  /** Toast msg */
  @Prop() msg$: Observable<Toast>

  @State() open = false
  @State() text = ''
  @State() status = ''

  /** Input Event */
  @Event({ composed: false }) state: EventEmitter

  private sub = new Subscription()
  private active$ = new Subject<boolean>()

  public componentWillLoad(): void {
    this.state.emit(false)
    this.sub.add(
      this.msg$?.subscribe(({ text, status }) => {
        this.status = status
        this.text = text
        this.active$.next(true)
      })
    )

    this.sub.add(
      this.active$
        .pipe(
          filter((state: boolean) => state),
          delay(10),
          tap(() => {
            this.open = true
            this.state.emit(true)
          }),
          delay(1500),
          tap(() => {
            this.open = false
            this.state.emit(false)
          }),
          delay(500)
        )
        .subscribe(() => {
          this.status = ''
          this.text = ''
          this.active$.next(false)
        })
    )
  }

  public disconnectedCallback(): void {
    this.sub.unsubscribe()
  }

  public render(): HTMLTfToastElement {
    return (
      <Host>
        {this.text !== '' && (
          <div class="toast__wrapper">
            <div
              class={`toast__msg toast__msg--${this.status}${this.open ? ' toast__msg--open' : ''}`}
            >
              {this.text}
            </div>
          </div>
        )}
      </Host>
    )
  }
}
