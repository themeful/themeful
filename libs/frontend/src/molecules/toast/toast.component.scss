:host {
  display: block;
}

.toast {
  &__wrapper {
    box-shadow: 0px -4px 4px -3px rgba(30, 30, 30, 0.2);
    margin-left: calc(50vw - 150px);
    width: 300px;
    position: absolute;
    overflow: hidden;
    height: 30px;
    top: 0;
    pointer-events: none;
    display: flex;
    justify-content: center;
  }
  &__msg {
    position: absolute;
    color: #fff;
    padding: 4px 8px;
    text-align: center;
    border-bottom-left-radius: 4px;
    border-bottom-right-radius: 4px;
    transition: opacity 300ms ease-in-out, top 300ms ease-in-out;
    opacity: 0;
    top: -30px;
    min-width: 100px;

    &--error {
      background-color: rgba(224, 73, 62, 0.8);
    }

    &--success {
      background-color: rgba(95, 192, 17, 0.8);
    }

    &--open {
      top: 0;
      opacity: 1;
    }
  }
}
