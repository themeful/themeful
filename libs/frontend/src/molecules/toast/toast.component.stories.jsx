import { h } from '@stencil/core'

export default {
  title: 'Molecules/Toast',
  args: {
    first: 'John',
    middle: 'S',
    last: 'Doe',
  },
}

export const toast = (args) => {
  return <tf-toast {...args}></tf-toast>
}
