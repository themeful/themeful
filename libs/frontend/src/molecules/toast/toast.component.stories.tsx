import { h } from '@stencil/core'
import { interval, mapTo } from 'rxjs'

export default {
  title: 'Molecules/Toast',
}

export const toast = ({ first, second }): HTMLElement => {
  return (
    <div>
      <div class="header">
        <h1>Toast</h1>
      </div>
      <div class="content">
        <div style={{ position: 'relative', width: '400px', height: '200px' }}>
          <tf-toast {...first}></tf-toast>
        </div>
        <div style={{ position: 'relative', width: '400px', height: '200px' }}>
          <tf-toast {...second}></tf-toast>
        </div>
      </div>
    </div>
  )
}

toast.args = {
  first: {
    msg$: interval(5000).pipe(
      mapTo({
        text: 'Hello World',
        status: 'success',
      })
    ),
  },
  second: {
    msg$: interval(5000).pipe(
      mapTo({
        text: 'Hello World',
        status: 'error',
      })
    ),
  },
}
