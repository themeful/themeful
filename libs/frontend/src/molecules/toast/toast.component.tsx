import { Component, h, Prop } from '@stencil/core'

@Component({
  tag: 'tf-toast',
  styleUrl: 'toast.component.scss',
  shadow: true,
})
export class ToastComponent {
  /**
   * The first name
   */
  @Prop() first: string

  /**
   * The middle name
   */
  @Prop() middle: string

  /**
   * The last name
   */
  @Prop() last: string

  private getText(): string {
    return `${this.first} ${this.middle} ${this.last}`
  }

  render(): HTMLTfToastElement {
    return <div>Hello, World! I'm {this.getText()}</div>
  }
}
