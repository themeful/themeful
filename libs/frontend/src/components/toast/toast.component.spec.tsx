import { h } from '@stencil/core'
import { newSpecPage } from '@stencil/core/testing'
import { Toast } from '@typings'
import { Subject } from 'rxjs'
import { ToastComponent } from './toast.component'

describe('ToastComponent', () => {
  it('renders with values', async () => {
    const subject = new Subject<Toast>()
    const page = await newSpecPage({
      components: [ToastComponent],
      autoApplyChanges: true,
      template: () => <tf-toast msg$={subject} />,
    })
    subject.next({
      text: 'Hello World',
      status: 'success',
    })
    subject.complete()
    expect(page.root).toMatchSnapshot()
  })
})
