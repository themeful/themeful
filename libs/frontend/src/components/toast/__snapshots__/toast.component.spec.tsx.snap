// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastComponent renders with values 1`] = `
<tf-toast>
  <mock:shadow-root>
    <div class="toast__wrapper">
      <div class="toast__msg toast__msg--open toast__msg--success">
        Hello World
      </div>
    </div>
  </mock:shadow-root>
</tf-toast>
`;
